src/host.rs

```diff
@@ -55,10 +57,20 @@ impl Host {
         } else {
             let decoded = percent_decode(input.as_bytes());
             let domain = String::from_utf8_lossy(&decoded);
-            // TODO: Remove this check and use IDNA "domain to ASCII"
+
+            let mut domain = match idna::domain_to_ascii(&domain) {
+                Ok(s) => s,
+                Err(_) => return Err(ParseError::InvalidDomainCharacter)
+            };
+
             if !domain.is_ascii() {
-                Err(ParseError::NonAsciiDomainsNotSupportedYet)
-            } else if domain.find(&[
+                domain = match punycode::encode_str(&domain) {
```

Comment:
Converting to punycode should be part of `domain_to_ascii` per http://www.unicode.org/reports/tr46/#ToASCII , and should happen on individual dot-separated labels, not the entire domain.